aptos-move/aptos-aggregator/src/delta_change_set.rs

```diff
@@ -6,11 +6,79 @@
 //! postcondition.
 
 use crate::{
-    bounded_math::SignedU128,
+    bounded_math::{SignedU128, BoundedMath},
     delta_math::{merge_data_and_delta, merge_two_deltas, DeltaHistory},
-    types::{code_invariant_error, DelayedFieldsSpeculativeError, PanicOr},
+    types::{code_invariant_error, DelayedFieldsSpeculativeError, PanicOr, DeltaApplicationFailureReason},
 };
 
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub struct DeltaWithMax {
+    /// Delta which is the result of the execution.
+    pub update: SignedU128,
+    /// The maximum value the aggregator can reach.
+    pub max_value: u128,
+}
+
+impl DeltaWithMax {
+    /// Creates a new delta op.
+    pub fn new(update: SignedU128, max_value: u128) -> Self {
+        Self {
+            max_value,
+            update,
+        }
+    }
+
+    /// Returns the kind of update for the delta op.
+    pub fn get_update(&self) -> SignedU128 {
+        self.update
+    }
+
+    /// Returns the result of delta application to `base` or error if
+    /// postcondition is not satisfied.
+    pub fn apply_to(&self, base: u128) -> Result<u128, DelayedFieldsSpeculativeError> {
+        let math = BoundedMath::new(self.max_value);
+        match self.update {
+            SignedU128::Positive(value) => math.unsigned_add(base, value).map_err(|_e| DelayedFieldsSpeculativeError::DeltaApplication {
+                base_value: base,
+                max_value: self.max_value,
+                delta: self.update,
+                reason: DeltaApplicationFailureReason::Overflow,
+            }),
+            SignedU128::Negative(value) => math.unsigned_subtract(base, value).map_err(|_e| DelayedFieldsSpeculativeError::DeltaApplication {
+                base_value: base,
+                max_value: self.max_value,
+                delta: self.update,
+                reason: DeltaApplicationFailureReason::Underflow,
+            }),
+        }
+    }
+
+    pub fn create_merged_delta(
+        prev_delta: &DeltaWithMax,
+        next_delta: &DeltaWithMax,
+    ) -> Result<DeltaWithMax, PanicOr<DelayedFieldsSpeculativeError>> {
+        if prev_delta.max_value != next_delta.max_value {
+            Err(code_invariant_error(
+                "Cannot merge deltas with different limits",
+            ))?;
+        }
+
+        let new_delta = BoundedMath::new(prev_delta.max_value).signed_add(&prev_delta.update, &next_delta.update).map_err(|_| DelayedFieldsSpeculativeError::DeltaMerge {
+            base_delta: prev_delta.update,
+            delta: next_delta.update,
+            max_value: prev_delta.max_value,
+        })?;
+
+        Ok(DeltaWithMax::new(new_delta, prev_delta.max_value))
+    }
+
+    pub fn into_op_no_additional_history(self) -> DeltaOp {
```

Comment:
I feel like having no duplication and explicit enum for history / no history is cleaner?

Reply:
I feel the same. We can just have a DeltaOp with `history: Option<DeltaHistory>`?

Reply:
let's figure that out separately from this fix and this PR.

But I think it is more likely we should delete DeltaOp, than have Option<DeltaHistory>.


Alternatively, we can have:
```
  DeltaOp {
    delta: DeltaWithMax,
    history: DeltaHistory,
  }
```

which seems better than Option<DeltaHistory>. 

In general DeltaOp is used in V1, and I don't want to modify V1 path, unless necessary, before v1.8